src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp

```diff
@@ -0,0 +1,401 @@
+	static constexpr double mean_molecular_weight = NAN;
+	static constexpr double boltzmann_constant = quokka::boltzmann_constant_cgs;
+};
+
+template <> struct Physics_Traits<ShocktubeProblem> {
+	// cell-centred
+	static constexpr bool is_hydro_enabled = true;
+	static constexpr bool is_chemistry_enabled = false;
+	static constexpr int numMassScalars = 3;		     // number of mass scalars
+	static constexpr int numPassiveScalars = numMassScalars + 0; // number of passive scalars
+	static constexpr bool is_radiation_enabled = false;
+	// face-centred
+	static constexpr bool is_mhd_enabled = false;
+};
+
+// left- and right- side shock states
+constexpr amrex::Real rho_L = 1.0;
+constexpr amrex::Real P_L = 1.0;
+constexpr amrex::Real rho_R = 0.125;
+constexpr amrex::Real P_R = 0.1;
+
+template <> void RadhydroSimulation<ShocktubeProblem>::setInitialConditionsOnGrid(quokka::grid grid_elem)
+{
+	// extract variables required from the geom object
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx = grid_elem.dx_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = grid_elem.prob_lo_;
+	const amrex::Box &indexRange = grid_elem.indexRange_;
+	const amrex::Array4<double> &state_cc = grid_elem.array_;
+
+	const int ncomp_cc = Physics_Indices<ShocktubeProblem>::nvarTotal_cc;
+	// loop over the grid and set the initial condition
+	amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+		amrex::Real const x = prob_lo[0] + (i + amrex::Real(0.5)) * dx[0];
+
+		const double vx = 0.0;
+		double rho = NAN;
+		double P = NAN;
+		std::array<Real, 3> specie = {-1.0}; // why can't this loop read in numPassiveScalars
+
+		// from Plewa and Muller 1999
+		if (x <= 0.5) {
+			rho = rho_L;
+			P = P_L;
+		} else {
+			rho = rho_R;
+			P = P_R;
+		}
+
+		// initialize mass fractions of species
+		if (x <= 0.5) {
+			specie[0] = 0.8;
+		} else if (x > 0.5 && x <= 0.75) {
+			specie[0] = 0.3;
+		} else {
+			specie[0] = 0.1;
+		}
+
+		specie[1] = 0.3 * pow(sin(20 * 3.14 * x), 2);
+		specie[2] = 1 - specie[0] - specie[1];
+
+		AMREX_ASSERT(!std::isnan(vx));
+		AMREX_ASSERT(!std::isnan(rho));
+		AMREX_ASSERT(!std::isnan(P));
+
+		const auto gamma = quokka::EOS_Traits<ShocktubeProblem>::gamma;
+		for (int n = 0; n < ncomp_cc; ++n) {
+			state_cc(i, j, k, n) = 0.;
+		}
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::density_index) = rho;
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::x1Momentum_index) = rho * vx;
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::x2Momentum_index) = 0.;
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::x3Momentum_index) = 0.;
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::energy_index) = P / (gamma - 1.) + 0.5 * rho * (vx * vx);
+		state_cc(i, j, k, HydroSystem<ShocktubeProblem>::internalEnergy_index) = P / (gamma - 1.);
+
+		for (int nn = 0; nn < 3; ++nn) {
+			state_cc(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + nn) = specie[nn];
+		}
+	});
+}
+
+template <>
```

Comment:
I haven't changed anything down below (so it's identical to `HydroShocktube` test file). How do I add the scalars in the code from here on?

Reply:
So I don't need to set anything for species in `setCustomBoundaryConditions`?

Reply:
I think if the conservation errors generated by Quokka after the test runs are ~ machine precision level, that would mean CMA is working? I don't think this test has an exact solution: the original paper compares the error in mass conservation between different schemes to show CMA performs the best, and returns conservation errors of the order of 1e-14.

Reply:
That's a good point. We don't want to use `ComputeReferenceSolution` then, since that only works if you have a reference solutiont to compare to. You'll need to move the code to extract the numerical solution to a vector to the after `sim.evolve()` where you can implement this check.

Reply:
I think you should do it at the end of the simulation. Yes, it should work in all dimensions.